docs: Add the NavManager guide (#1023)
AmadeusITGroup/AgnosUI: docs/04-Services/04-Navigation-Manager.md

---
since: 0.1.1
---

# Navigation Manager

## Service Overview

This service simplifies keyboard navigation for a set of DOM elements. It exposes a factory function `createNavManager()` that creates an object containing:

- `elementsInDomOrder$` - a store containing the navigable elements in DOM order
- `directive` - a directive to use on some DOM elements, both to add the keydown event handler and to specify which elements should be managed (either by directly putting the directive on those elements, or by putting the directive on a parent element and specifying which child elements should be included through a `selector` function via `NavManagerItemConfig`)
- Utilities to navigate the items:
  - `refreshElements` - method to manually refresh the `elementsInDomOrder$` to capture the latest state of DOM
  - `focusIndex` - method to focus the element at the given index within `elementsInDomOrder$`

>   - `focusPrevious` - method to focus the previous element, respecting the anscestor direction

Reviewer: There is a typo with the `ancestor` word (also in the following lines).
I think it is not clear which ancestor we are talking about. I know from the code, but it is not explained in the doc that it is the closest common ancestor of all elements which is considered.

Reviewer: Maybe it should be explained that `focusPrevious`, `focusNext`, `focusFirst` and `focusLast` refer to the DOM order, and that `focusLeft`, `focusRight`, `focusFirstLeft` and `focusFirstRight` will be do exactly the same in a LTR configuration and the opposite in a RTL configuration.

  - `focusLast` - method to focus the last element in the DOM order
  - `focusPrevious` - method to focus the previous element in the DOM order
  - `focusNext` - method to focus the next element in the DOM order

>   - `focusFirstLeft` - method to focus the element at the left-end of the list and will do the opposite in a RTL configuration

Reviewer: For me the sentence is misleading regarding "the opposite". The opposite of "left-end" ? Does it mean "right-end" ? (of course not !)
I would prefer something like this:
```suggestion
  - `focusFirstLeft` - method to focus the element at the left-end of the list (identical to `focusFirst` in a LTR configuration and to `focusLast` in a RTL configuration)
```
(same remark for `focusFirstRight`, `focusLeft` and `focusRight`)

Author: makes sense :) fixed!

  - `focusLast` - method to focus the last element, respecting the anscestor direction
  - `focusLeft` - method to focus the next focusable element to the left of the currently focused element
  - `focusRight` - method to focus the next focusable element to the right of the currently focused element

## Usage

Whenever you want to custom keyboard navigation within a defined set of DOM elements you can simply:

- use the factory function to create the NavManager object
- use the `NavManagerItemConfig` to configure the NavManger by setting:

>   - `keys` - a map of key handlers. The key in the map should match the key event (for example, "ArrowLeft" or "Ctrl+PageDown"). The value in the map is the corresponding key handler. (for example, "FocusNext" or "FocusPrevious")

Reviewer: ```suggestion
  - `keys` - a map of key handlers. The key in the map should match the key event (for example, `"ArrowLeft"` or `"Ctrl+PageDown"`). The value in the map is the corresponding key handler (for example, the `focusNext` or `focusPrevious` function returned by `createNavManager`)
```
(so that people don't think the value is a string)
